File: 872/main.go
package main

import (
	"fmt"
	"reflect"
)

func main() {
	tree1 := &TreeNode{
		1,
		nil,
		nil,
	}
	tree2 := &TreeNode{
		1,
		&TreeNode{
			2,
			nil,
			nil,
		},
		nil,
	}
	fmt.Println(leafSimilar(tree1, tree2))
}

// Definition for a binary tree node.
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func leafSimilar(root1 *TreeNode, root2 *TreeNode) bool {
	return reflect.DeepEqual(getLeafSeq(root1), getLeafSeq(root2))
}

func getLeafSeq(tree *TreeNode) []int {
	result := []int{}
	if tree == nil {
		return result
	} else if tree.Left == nil && tree.Right == nil {
		result = append(result, tree.Val)
	} else {
		result = append(getLeafSeq(tree.Left), getLeafSeq(tree.Right)...)
	}

	return result
}

File: 872/main.py
from typing import Optional


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def leafSimilar(self, root1: Optional[TreeNode], root2: Optional[TreeNode]) -> bool:
        """
        approach
        get leaf sequence
        compare
        get leaf sequence is easy, each recursive call should return [val] if leaf, otherwise concat left and right recursive lists
        """

        def leaf_sequence_recur(root: Optional[TreeNode]):
            if root is None:
                return []
            if root.left is None and root.right is None:
                return [root.val]
            left = leaf_sequence_recur(root.left)
            right = leaf_sequence_recur(root.right)
            return left + right

        a = leaf_sequence_recur(root1)
        b = leaf_sequence_recur(root2)
        if len(a) != len(b):
            return False
        for x, y in zip(a, b):
            if x != y:
                return False
        return True
